drivers/ethernet/eth_xmc4xxx.c

```diff
@@ -0,0 +1,1144 @@
+		/* reinitialize the descriptors and set dma_desc_tx_head to 0 */
+
+		dma_desc = &tx_dma_desc[dev_data->dma_desc_tx_head];
+
+		frag_data = frag->data;
+		frag_len = frag->len;
+
+		dma_desc->buffer1 = (volatile uint32_t)frag_data;
+		dma_desc->length = frag_len;
+
+		/* give ownership of descriptor back to dma and set checksum offload */
+		/* TCH we are using a circular list */
+		dma_desc->status = ETH_MAC_DMA_TDES0_CIC | ETH_MAC_DMA_TDES0_TCH;
+
+		if (!first_descriptor) {
+			/* Delay giving ownership of first frag to DMA. Prevents race condition */
+			/* where second other frags are not ready */
+			dma_desc->status |= ETH_MAC_DMA_TDES0_OWN;
+		} else {
+			dma_desc->status |= ETH_MAC_DMA_TDES0_FS;
+
+#if defined(CONFIG_NET_GPTP)
+			struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+
+			if (ntohs(hdr->type) == NET_ETH_PTYPE_PTP) {
+				dma_desc->status |= ETH_MAC_DMA_TDES0_TTSE;
+			}
+#endif
+		}
+		first_descriptor = false;
+
+		tx_frame->head_index = dev_data->dma_desc_tx_head;
+
+		MODULO_INC_TX(dev_data->dma_desc_tx_head);
+
+		irq_unlock(key);
+
+		frag = frag->frags;
+	}
+
+	unsigned int key = irq_lock();
+
+	if (dev_data->tx_frames_flushed) {
+		k_mem_slab_free(&tx_frame_slab, (void **)&tx_frame);
+		net_pkt_unref(pkt);
+		irq_unlock(key);
+		LOG_WRN("Flushing tx frames");
+		return -EIO;
+	}
+
+	/* label last dma descriptor as last segment and trigger interrupt on last segment */
+	dma_desc->status |= ETH_MAC_DMA_TDES0_IC | ETH_MAC_DMA_TDES0_LS;
+
+	/* Finally give ownership of first frag to DMA. After this point the DMA engine */
+	/* may transfer the whole frame from RAM to Ethernet */
+	tx_dma_desc[tx_frame->tail_index].status |= ETH_MAC_DMA_TDES0_OWN;
+
+	sys_slist_append(&dev_data->tx_frame_list, &tx_frame->node);
+
+	eth_xmc4xxx_trigger_dma_tx(dev_cfg->regs);
+
+	irq_unlock(key);
+
+	return ret;
+}
+
+static struct net_pkt *eth_xmc4xxx_rx_pkt(const struct device *dev)
+{
+	struct eth_xmc4xxx_data *dev_data = dev->data;
+	const struct eth_xmc4xxx_config *dev_cfg = dev->config;
+	struct net_pkt *pkt = NULL;
+	struct net_buf *new_frag;
+
+	bool eof_found = false;
+	uint16_t tail;
+	XMC_ETH_MAC_DMA_DESC_t *dma_desc;
+	int num_frags = 0;
+	uint16_t frame_end_index;
+	struct net_buf *frag, *last_frag = NULL;
+
+	tail = dev_data->dma_desc_rx_tail;
+	dma_desc = &rx_dma_desc[tail];
+
+	if (IS_OWNED_BY_DMA_RX(dma_desc)) {
+		return NULL;
+	}
+
+	if (!IS_START_OF_FRAME_RX(dma_desc)) {
+		/* handle this error - missing SOF packet? */
+		eth_xmc4xxx_flush_rx(dev);
+		return NULL;
+	}
+
+	while (!IS_OWNED_BY_DMA_RX(dma_desc)) {
+		eof_found = IS_END_OF_FRAME_RX(dma_desc);
+		num_frags++;
+		if (eof_found) {
+			break;
+		}
+
+		MODULO_INC_RX(tail);
+
+		if (tail == dev_data->dma_desc_rx_tail) {
+			/* wrapped */
+			break;
+		}
+
+		dma_desc = &rx_dma_desc[tail];
+	}
+
+	if (!eof_found) {
+		return NULL;
+	}
+
+	frame_end_index = tail;
+
+	pkt = net_pkt_rx_alloc(K_NO_WAIT);
+	if (pkt == NULL) {
+		LOG_ERR("Net packet allocation error");
```

Comment:
It's a good point, I'll update the log level to DBG and update stats.

An error message that I think is useful is 
LOG_ERR("Frag allocation error. Increase CONFIG_NET_BUF_RX_COUNT."); on eth_xmc4xxx.c:432

Because the driver pre-allocates buffers for the DMA, it's easy to have too few buffers configured with CONFIG_NET_BUF_RX_COUNT.
The error message would be useful to quickly find the problem. Perhaps I could change this log message so that it's only printed once. Let me know what you think.